staging/src/k8s.io/apiserver/pkg/endpoints/handlers/audit.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handlers
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+)
+
+func auditRequest(ae *audit.Event, obj runtime.Object, scope RequestScope) {
+	if ae == nil || ae.Level.Less(audit.LevelRequest) {
+		return
+	}
+
+	// TODO: hook into the serializer to avoid double conversion
+	jsonObj, err := responsewriters.EncodeAuditObject(obj, scope.Resource.GroupVersion(), scope.Serializer)
+	if err != nil {
+		// TODO: add error slice to audit event struct
+		glog.Warningf("Auditing failed of %q request: %v", scope.Resource, err)
+		return
+	}
+	ae.RequestBody = jsonObj
+
+	// complete ObjectRef
+	if ae.ObjectRef == nil {
+		ae.ObjectRef = &audit.ObjectReference{}
+	}
+	if len(ae.ObjectRef.Kind) == 0 {
+		ae.ObjectRef.Kind = obj.GetObjectKind().GroupVersionKind().Kind
+	}
+	if acc, ok := obj.(v1.ObjectMetaAccessor); ok {
+		meta := acc.GetObjectMeta()
+		if len(ae.ObjectRef.Namespace) == 0 {
```

Comment:
Seems like namespace and name should come off of the request information, not the object.